cluster-autoscaler/proposals/resourceslack.md

```diff
@@ -0,0 +1,235 @@
+# Resource Slack
+
+## Abstract
+
+"Resource slack" is some extra capacity in a k8s cluster kept for upcoming pods so that a newly created pod fit within resource slack does not need to wait until an extra node is started and gets ready.
+It enables pods to be scaled out faster while minimizing impact to resource utilization.
+This document sketches how the resource slack feature will be implemented in a future release of K8S.
+
+## Motivation
+
+Currently, ClusterAutoscaler tries to maintain necessary and sufficient amount of nodes.
+For scaling up, it adds nodes only after some unschedulable pods are observed.
+This behavior of scaling up implies that, in a worst case, an unschedulable pod must wait until an eligible node is created, started, and then gets ready before it is finally scheduled.
+A node takes a much longer time than that of a pod to start and become ready and therefore it results in pods to occasionally delayed for several minutes to be horizontally scaled out.
+A possible work-around can be done today is to lower the threshold of HPA to trigger a scaling up of a replicaset/deployment earlier but it reduces total resource utilization when there are many replicasets and deployments required the work-around.
+Having the resource slack feature allows us to scale out pods faster while minimizing impact to resource utilization.
+
+## Design
+
+To achieve Resource Slack, introduce placeholder pods that will live only inside of ClusterAutoscaler.
+They are created to virtually request certain amounts of CPU and memory so that they influence all of the scale-up/scale-down decision made in CA, but won't be present in the scheduler running inside an actual K8S cluster, so the new pods will be able to pick the place of the placeholders and push the placeholders elsewhere (and make them trigger the scale-up/down).
+
+### Requirements
+
+Version 1
+
+* [R1] Works on relatively large clusters today's CA used to work (dozens of nodes, hundreds of pods? Correct me if necessary!)
+  * In general we don't want to have O(pods * nodes) operations every single loop
+* [R2] Keep the specified amount of extra capacity at minimum 
+* [R3] Ensure full compatibility with scale down
+* [R4] Only expose necessary and sufficient knobs to users
+
+Version 2
+
+* [R5] Don't add unnecessarily large node
+* [R6] Provide extra space even for the "biggest" pod in term of memory and cpu
+* [R7] Provide extra space even for every pod which is not biggest but do request some cpu(< max(pod_request_cpu)) and memory(< max(pod_request_mmory))
+  * Say, the "biggest pods" request 1000m cpu/1G mem, 10m cpu/10G mem respectively
+  * Even if they could fit within existing free space, a 500m(<1000m cpu)/5G mem pod has chances not to fit any node 
+
+Version 3
+
+* [R8] Reserve space even for the pods having tolerations
+* [R9] Reserve space even for the pods having nodeSelector/nodeAffinity
```

Comment:
You meant to annotate k8s objects to explicitly mark them so that they're taken into account for extra capacity, regardless of a pod has nodeSelector/nodeAffinity/podAntiAffinity or not?